include/bamit/IntervalNode.hpp

```diff
@@ -111,146 +149,216 @@ class IntervalNode
+void construct_tree(std::unique_ptr<IntervalNode> & node,
+                    std::vector<std::vector<Record>> & records_i,
+                    int32_t const offset = 0)
 {
-    if (records_i.empty())
+    // If there are no records, exit.
+    if (records_i.empty() || (records_i.front().empty() && records_i.back().empty()))
     {
         return;
     }
     // Set node to an empty IntervalNode pointer.
     node = std::make_unique<IntervalNode>();
 
-    // Calculate and set median.
-    node->set_median(calculate_median(records_i));
-    Position cur_median = node->get_median();
+    // Calculate and set chromosome.
+    auto [cur_chr, cur_median] = calculate_median(records_i);
+    node->set_chr(cur_chr + offset);
 
     // Get reads which intersect median.
-    std::vector<Record> lRecords{};
-    std::vector<Record> rRecords{};
-    for (auto & r : records_i)
+    std::vector<std::vector<Record>> lRecords{};
+    std::vector<std::vector<Record>> rRecords{};
+
+    // Add initial empty vector to rRecords.
+    rRecords.emplace_back();
+    for (size_t i = 0; i < records_i.size(); ++i)
     {
-        // Median is to the left of the read, so read is in right subtree.
-        if (cur_median < r.start)
+        // Chromosome is less than current chromosome.
+        if (i < cur_chr)
         {
-            rRecords.push_back(std::move(r));
+            lRecords.push_back(std::move(records_i[i]));
         }
-        // Median is to the right of the read, so read is in left subtree.
-        else if (cur_median > r.end)
+        // Chromosome is greater than current chromosome.
+        else if (i > cur_chr)
         {
-            lRecords.push_back(std::move(r));
+            rRecords.push_back(std::move(records_i[i]));
         }
-        // If median is not to the left or right, it must be within the read.
-        // Set the file_offset for this node only if it has not yet been set.
-        else if (node->get_file_offset() == -1)
+        // We are at current chromosome. Reads must be added to current node if they intersect interval.
+        // Otherwise, they should be added to the left/right trees.
+        else
         {
-            node->set_file_offset(r.file_offset);
+            std::vector<Record> l_cur_chr{};
+            uint32_t start{0}, end{0};
+            for (auto & r : records_i[i])
+            {
+                // Read ends before the median.
+                if (r.end < cur_median)
+                {
+                    // Push it back into the first empty vector.
+                    l_cur_chr.push_back(std::move(r));
+                }
+                // Read starts after the median.
+                else if (r.start > cur_median)
+                {
+                    rRecords[0].push_back(std::move(r));
+                }
+                // Read intersects the median. Only store file offset and start from the left-most read!
+                // End is always updated while the read intersects the median.
```

Comment:
Why? Then, we don't have the rightmost end, do we?

Reply:
We do because reads are sorted in ascending order!